cmd/virt-tail/main.go

```diff
@@ -0,0 +1,228 @@
+	socketExists := checkFile(socketFile)
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		log.Log.V(3).Infof("watcher error: %v", err)
+		return err
+	}
+	defer watcher.Close()
+
+	// Add a path.
+	dirPath := filepath.Dir(gLogFile)
+	found := false
+	i := 0
+	for i < 30 && !found {
+		i = 1 + 1
+		if _, derr := os.Stat(dirPath); derr == nil {
+			found = true
+			if err = watcher.Add(dirPath); err != nil {
+				log.Log.V(3).Infof("watcher error: %v - %s", err, dirPath)
+				return err
+			}
+		} else {
+			time.Sleep(100 * time.Millisecond)
+		}
+	}
+	if !found {
+		rerr := errors.New("expected directory is still not ready")
+		log.Log.V(3).Infof("watchFS error: %v", rerr)
+		return rerr
+	}
+
+	// initial timeout for serial console socket creation
+	socketCheckCh := make(chan int)
+	time.AfterFunc(initialSocketTimeout, func() {
+		socketCheckCh <- 1
+	})
+
+	// Start listening for events.
+	for {
+		select {
+		case <-socketCheckCh:
+			if !socketExists {
+				if socketExists = checkFile(socketFile); !socketExists {
+					rerr := errors.New("socketFile is still not ready")
+					log.Log.V(3).Infof("watchFS error: %v", rerr)
+					return rerr
+				}
+			}
+			if checkFile(termFile) {
+				log.Log.V(3).Infof("watchFS error: termFile was already there")
+				return &TermFileError{}
+			}
+		case event := <-watcher.Events:
+			if event.Has(fsnotify.Create) {
+				if event.Name == termFile {
+					// termination file got created, we should quickly terminate
+					terr := &TermFileError{}
+					log.Log.V(3).Infof("watchFS error: %v", terr)
+					return terr
+				} else if event.Name == socketFile {
+					// socket file got created
+					socketExists = true
+				}
+			} else if event.Has(fsnotify.Remove) {
+				if event.Name == socketFile {
+					// socket file got deleted, we should quickly terminate
+					rerr := errors.New("socketFile got removed")
+					log.Log.V(3).Infof("watchFS error: %v", rerr)
+					return rerr
+				}
+			}
+		case werr := <-watcher.Errors:
+			log.Log.V(3).Infof("watcher error: %v", werr)
+			return werr
+		case <-gctx.Done():
+			return gctx.Err()
+		}
+	}
+}
+
+func main() {
+	// set new default verbosity, was set to 0 by glog
+	goflag.Set("v", "2")
+	pflag.CommandLine.AddGoFlag(goflag.CommandLine.Lookup("v"))
+	pflag.CommandLine.ParseErrorsWhitelist = pflag.ParseErrorsWhitelist{UnknownFlags: true}
+	pflag.Parse()
+
+	log.InitializeLogging("virt-tail")
+	setTailLogverbosity()
+
+	logFile := pflag.String("logfile", "", "path of the logfile to be streamed")
+
+	pflag.Parse()
+
+	if logFile == nil || *logFile == "" {
+		log.Log.V(3).Infof("logfile flags must be provided")
+		os.Exit(1)
+	}
+	gLogFile = *logFile
+
+	// Create context that listens for the interrupt signal from the container runtime.
+	ctx, _ := signal.NotifyContext(context.Background(), os.Interrupt)
+
+	g, gctx = errgroup.WithContext(ctx)
+	g.Go(tailLogs)
+	g.Go(watchFS)
+
+	// wait for all errgroup goroutines
+	if err := g.Wait(); err != nil {
```

Comment:
Can it return `err == nil`?

Reply:
Get rid of explicit `os.Exit(0)` and return cleanly by default if `err != nil` but is `context.Canceled` or `TermFileError` then?

Reply:
done